File: examples/python/basic_room.py
#!/usr/bin/env python3
"""
Digital Samba - Basic Room Management (Python)

Demonstrates:
- Creating a room
- Generating participant tokens
- Listing rooms
- Deleting a room

Usage:
    pip install requests pyjwt
    DS_DEVELOPER_KEY=your-key DS_TEAM_ID=your-team-id python basic_room.py
"""

import os
import time
import requests
import jwt

# Configuration
DEVELOPER_KEY = os.environ.get('DS_DEVELOPER_KEY')
TEAM_ID = os.environ.get('DS_TEAM_ID')
BASE_URL = 'https://api.digitalsamba.com/api/v1'

if not DEVELOPER_KEY:
    print('Error: DS_DEVELOPER_KEY environment variable required')
    exit(1)


def api_call(method: str, endpoint: str, data: dict = None) -> dict:
    """Make an authenticated API call."""
    url = f'{BASE_URL}{endpoint}'
    headers = {
        'Authorization': f'Bearer {DEVELOPER_KEY}',
        'Content-Type': 'application/json'
    }

    response = requests.request(method, url, headers=headers, json=data)

    if response.status_code == 204:
        return None
    elif response.ok:
        return response.json()
    else:
        raise Exception(f'API Error {response.status_code}: {response.text}')


def create_room_token(room_id: str, user_name: str, user_id: str = None, role: str = 'attendee') -> str:
    """Generate a JWT token for room access (alternative to API token endpoint)."""
    if not TEAM_ID:
        raise ValueError('DS_TEAM_ID required for local JWT generation')

    payload = {
        'td': TEAM_ID,
        'rd': room_id,
        'u': user_name,
        'ud': user_id,
        'role': role,
        'exp': int(time.time()) + 3600  # 1 hour
    }

    return jwt.encode(payload, DEVELOPER_KEY, algorithm='HS256')


def main():
    try:
        # 1. Create a room
        print('1. Creating room...')
        room = api_call('POST', '/rooms', {
            'friendly_url': f'demo-room-{int(time.time())}',
            'privacy': 'public',
            'description': 'Demo room created via Python',
            'max_participants': 50,
            'chat_enabled': True,
            'recordings_enabled': True
        })
        print(f'   Room created: {room["room_url"]}')
        print(f'   Room ID: {room["id"]}')

        # 2a. Generate token via API
        print('\n2a. Generating token via API...')
        token_response = api_call('POST', f'/rooms/{room["id"]}/token', {
            'ud': 'user-123',
            'u': 'Demo User',
            'role': 'moderator'
        })
        print(f'   Join link: {token_response["link"]}')

        # 2b. Generate token locally (alternative)
        if TEAM_ID:
            print('\n2b. Generating token locally...')
            local_token = create_room_token(
                room_id=room['id'],
                user_name='Local Token User',
                user_id='user-456',
                role='attendee'
            )
            print(f'   Local token generated (first 50 chars): {local_token[:50]}...')

        # 3. List rooms
        print('\n3. Listing rooms...')
        rooms = api_call('GET', '/rooms?limit=5')
        print(f'   Total rooms: {rooms["total_count"]}')
        for r in rooms['data'][:3]:
            print(f'   - {r["friendly_url"]} ({r["privacy"]})')

        # 4. Get room details
        print(f'\n4. Getting room details...')
        details = api_call('GET', f'/rooms/{room["id"]}')
        print(f'   Max participants: {details["max_participants"]}')
        print(f'   Chat enabled: {details["chat_enabled"]}')
        print(f'   Recordings enabled: {details["recordings_enabled"]}')

        # 5. Delete room
        print('\n5. Deleting room...')
        api_call('DELETE', f'/rooms/{room["id"]}')
        print('   Room deleted')

        print('\nDemo complete!')

    except Exception as e:
        print(f'Error: {e}')
        exit(1)


if __name__ == '__main__':
    main()

File: examples/node/basic-room.js
/**
 * Digital Samba - Basic Room Management (Node.js)
 *
 * Demonstrates:
 * - Creating a room
 * - Generating participant tokens
 * - Listing rooms
 * - Deleting a room
 *
 * Usage:
 *   DS_DEVELOPER_KEY=your-key DS_TEAM_ID=your-team-id node basic-room.js
 */

const https = require('https');

const DEVELOPER_KEY = process.env.DS_DEVELOPER_KEY;
const TEAM_ID = process.env.DS_TEAM_ID;

if (!DEVELOPER_KEY) {
  console.error('Error: DS_DEVELOPER_KEY environment variable required');
  process.exit(1);
}

// Helper function for API calls
function apiCall(method, path, data = null) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'api.digitalsamba.com',
      port: 443,
      path: path,
      method: method,
      headers: {
        'Authorization': `Bearer ${DEVELOPER_KEY}`,
        'Content-Type': 'application/json'
      }
    };

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        if (res.statusCode === 204) {
          resolve(null);
        } else if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(JSON.parse(body));
        } else {
          reject(new Error(`API Error ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', reject);

    if (data) {
      req.write(JSON.stringify(data));
    }
    req.end();
  });
}

async function main() {
  try {
    // 1. Create a room
    console.log('1. Creating room...');
    const room = await apiCall('POST', '/api/v1/rooms', {
      friendly_url: `demo-room-${Date.now()}`,
      privacy: 'public',
      description: 'Demo room created via Node.js',
      max_participants: 50,
      chat_enabled: true,
      recordings_enabled: true
    });
    console.log(`   Room created: ${room.room_url}`);
    console.log(`   Room ID: ${room.id}`);

    // 2. Generate a participant token
    console.log('\n2. Generating participant token...');
    const tokenResponse = await apiCall('POST', `/api/v1/rooms/${room.id}/token`, {
      ud: 'user-123',
      u: 'Demo User',
      role: 'moderator'
    });
    console.log(`   Token generated`);
    console.log(`   Join link: ${tokenResponse.link}`);

    // 3. List all rooms
    console.log('\n3. Listing rooms...');
    const rooms = await apiCall('GET', '/api/v1/rooms?limit=5');
    console.log(`   Total rooms: ${rooms.total_count}`);
    rooms.data.slice(0, 3).forEach(r => {
      console.log(`   - ${r.friendly_url} (${r.privacy})`);
    });

    // 4. Delete the room
    console.log('\n4. Deleting room...');
    await apiCall('DELETE', `/api/v1/rooms/${room.id}`);
    console.log('   Room deleted');

    console.log('\nDemo complete!');

  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();
